### PR TITLE
fix(browser): merge event-specific groups with registered groups

### LINKED ISSUE #1545: [Bug][Groups] It's not possible to opt out subsequent group event capture
Full context in this [slack thread](https://posthog.slack.com/archives/C0368RPHLQH/p1732198880119639?thread_ts=1732120428.677349&cid=C0368RPHLQH). Here's a summary:

1. Calling `posthog.group()` has a side-effect where subsequent events that captured are associated to the group.
2. This is fine if you have a high-level group type like `company`, since you probably want all the users events to be associated with the company.
3. However, for lower-level group types (e.g. `project` in an Asana-like app), you dont want ALL subsequent events to be associated with the group. Instead, you want to explicitly specify which events should be capture (see reason for this below)
4. @neilkakkar suggested to set the `groups` property in events to associate events e.g. `posthog.capture('new_task_created, { $groups: { project: <project_id> }`. **This doesnt work as expected** . See below for more
5. Also, if you try to set group properties by manually capturing`$groupidentify`, all subsequent events are still associated with the group.

e.g. if you call:
```
    posthog.capture('$groupidentify', {
      '$group_type': 'project',
      '$group_key': newProject.id,
      '$group_set': {
        name: newProjectName
      }
  });
```
In the above, all subsequent client-side events are then associated with the project.

## Why you might not want all subsequent events to be associated with the group

Consider the following scenario:
1. User has two tabs open. One on the homepage, one on a "create project" page
2. In create project page, user creates new project. So posthog.group() is called.
3. User does a bunch of events in the projects, which are correctly associated with the project.
4. User switches tab to dashboard. All events from the dashboard are now assocaited with the project 🤦🏻‍♂️
5. User switches back to projects tab. Events should be associated with projects again

## Why `posthog.capture('new_task_created, { $groups: { project: <project_id> }` doesnt work

Setting the $groups key will override any previous `posthog.group()` calls. For example:

1. Call `posthog.group('company', <company_id>)`
2. Then call `posthog.capture('new_task_created, { $groups: { project: <project_id> }`

**Expected result:**
Your event is captured with both the project and the company

**Actual result:**
Event is only captured with the project. The company is dropped

### PR DESCRIPTION
**Pull request by marandaneto**
## Problem

Calling `posthog.group()` stores a group for subsequent events. When a capture also supplies event-specific `$groups`, that object currently replaces every registered group. A project-specific event can therefore lose its registered company association.

Closes #1545.

## Changes

- Merge non-empty event-specific `$groups` with groups registered on the browser client.
- Let event-specific group keys take precedence without changing the registered groups.
- Preserve `$groups: {}` as a way to omit registered groups from one event.
- Add Vitest regression tests for merging, key precedence, persistence, and the empty-object behavior.

### Libraries affected

- [ ] All of them
- [x] posthog-js (web)
- [ ] posthog-js-lite (web lite)
- [ ] posthog-node
- [ ] posthog-react-native
- [ ] @posthog/react-native-plugin
- [ ] @posthog/react
- [ ] @posthog/ai
- [ ] @posthog/convex
- [ ] @posthog/next
- [ ] @posthog/nextjs-config
- [ ] @posthog/nuxt
- [ ] @posthog/openfeature-node-provider
- [ ] @posthog/openfeature-web-provider
- [ ] @posthog/rollup-plugin
- [ ] @posthog/webpack-plugin
- [ ] @posthog/types
- [ ] @posthog/browser-common

## Checklist

- [x] Tests for new code
- [x] Accounted for the impact of any changes across different platforms
- [x] Accounted for backwards compatibility of any changes (no breaking changes!)
- [x] Took care not to unnecessarily increase the bundle size

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Implemented with the Pi coding agent after reproducing the issue in an isolated worktree. The branch was merged with the Vitest migration, and the new tests were updated to use Vitest. Pi autoreview checked the final commit and reported no actionable findings.

### CODE REVIEW

File: packages/browser/src/__tests__/posthog-core-also.test.ts [hunk 1]
```diff
@@ -1741,6 +1741,46 @@ describe('posthog core', () => {
                     instance: 'app.posthog.com',
                 })
             })
+
+            it('merges event-specific groups with registered groups', () => {
+                posthog.group('company', 'company::5')
+                posthog.capture('some_event', { $groups: { project: 'project::7' } })
+
+                const eventPayload = vi.mocked(posthog._requestQueue!.enqueue).mock.calls[1][0]
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/browser/src/__tests__/posthog-core-also.test.ts:1749]: Fixed in fc2059eb3. Replaced the three direct array checks with the shared `isArray()` helper and removed their lint suppressions. The focused Vitest tests, full test file, typecheck, and lint all pass.

**Review by turnipdabeets — APPROVED** [review]
